vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -66,45 +92,47 @@
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[]
-        | select (.status.provisioning.state == "deprovisioning")
-        | .metadata.name ] | length'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: bmh_in_deprovisioning
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: saved_bmhs
     retries: 150
     delay: 10
-    until: bmh_in_deprovisioning.stdout == NUMBER_OF_BMH_MD
-
-  - name: Get the name of the deprovisioning BMH.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[]
-        | select (.status.provisioning.state == "deprovisioning")
-        | .metadata.name ] | add'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: deprovisioning_bmh_name
-
-  - name: Wait until above deprovisioning BMH is in Ready state again.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" | grep -w ready | awk '{print $1}'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: ready_bmh
+    until:
+      - saved_bmhs is succeeded
+      - saved_bmhs.resources | filter_provisioning("deprovisioning") | length == (NUMBER_OF_BMH_MD | int)
+
+  - name: Wait until above deprovisioning BMHs are in Ready state again.
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: bmhs
     retries: 200
     delay: 2
-    until: deprovisioning_bmh_name.stdout in ready_bmh.stdout_lines
+    until:
+      - bmhs is succeeded
+      - (saved_bmhs.resources | filter_provisioning("deprovisioning") | map(attribute='metadata.name')) |
+        intersect(bmhs.resources | filter_provisioning("ready") | map(attribute='metadata.name')) |
+        length ==  (NUMBER_OF_BMH_MD | int)
 
   - name: Check if just deprovisioned and became ready BMH is re-used for next provisioning.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" | grep -w provisioning | awk '{print $1}'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: provisioning_bmh
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: bmhs
     retries: 150
     delay: 2
-    until: deprovisioning_bmh_name.stdout in provisioning_bmh.stdout_lines
+    until:
+      - bmhs is succeeded
+      - (saved_bmhs.resources | filter_provisioning("deprovisioning") | map(attribute='metadata.name')) |
+        intersect(bmhs.resources | filter_provisioning("provisioning") | map(attribute='metadata.name')) |
+        length == (NUMBER_OF_BMH_MD | int)
```

Comment:
this should check if last deprovisioned but currently in `Ready` state BMH has been picked up or not, but you are filtering for `deprovisioning` BMH here? how are you making sure that just deprovisioned BMH in previous step has been reused?

Reply:
I am saving previous bmhs in `saved_bmhs` and current in `bmhs` then intersect deprovisioning from the **last** with provisioning from the **current** and ensure that all of the previous deprovisioning become in provisioning state. 
Is not that enough ?